### PR TITLE
Implement Google AuthServiceExtension

### PR DESCRIPTION
**Pull request by holachuy**
Issue #233   Migrating code from previous GoogleAuth to the new architecture. Implemented to TokensAndUrlAuthData type

### CODE REVIEW

File: extensions/auth/portability-auth-google/src/main/java/org/dataportabilityproject/auth/google/GoogleAuthDataGenerator.java [hunk 1]
```diff
@@ -0,0 +1,138 @@
...
+   * @param clientId The Application ID that the registration portal (apps.dev.microsoft.com)
+   *     assigned the portability instance
+   * @param clientSecret The application secret that was created in the app registration portal for
+   *     the portability instance
+   * @param httpTransport The http transport to use for underlying GoogleAuthorizationCodeFlow
+   * @param objectMapper The json factory provider
+   */
+  public GoogleAuthDataGenerator(
+      String redirectPath,
+      String clientId,
```
Comment (seehamrun) [inline extensions/auth/portability-auth-google/src/main/java/org/dataportabilityproject/auth/google/GoogleAuthDataGenerator.java]: This should take an AppCredential

File: settings.gradle [hunk 2]
```diff
@@ -12,6 +12,8 @@ include ':extensions:cloud:portability-cloud-local', ':extensions:cloud:portabil
 
 include ':extensions:auth:portability-auth-microsoft', ':extensions:auth:portability-auth-harness-microsoft'
 
+include ':extensions:auth:portability-auth-google'
```
Comment (seehamrun) [inline settings.gradle]: Lets also add this line to distributions/gateway-default/build.gradle

File: extensions/auth/portability-auth-google/src/main/java/org/dataportabilityproject/auth/google/GoogleAuthServiceExtension.java [hunk 3]
```diff
@@ -0,0 +1,92 @@
...
+    // TODO Create auth data generator for a given mode, usually means a different scope
+    if (DEFAULT_AUTH_GENERATOR_SERVICES.contains(transferDataType)) {
+      return authDataGenerator;
+    }
+    throw new UnsupportedOperationException();
+  }
+
+  @Override
+  public List<String> getImportTypes() {
+    return ImmutableList.of();
```
Comment (seehamrun) [inline extensions/auth/portability-auth-google/src/main/java/org/dataportabilityproject/auth/google/GoogleAuthServiceExtension.java]: Should this be DEFAULT_AUTH_GENERATOR_SERVICES? (here and in getExportTypes)

File: extensions/auth/portability-auth-google/src/main/java/org/dataportabilityproject/auth/google/GoogleAuthServiceExtension.java [hunk 4]
```diff
@@ -0,0 +1,92 @@
...
+
+  @Override
+  public List<String> getExportTypes() {
+    return ImmutableList.of();
+  }
+
+  @Override
+  public void initialize(ExtensionContext context) {
+
+    makeAuthDataGenerator(context);
```
Comment (seehamrun) [inline extensions/auth/portability-auth-google/src/main/java/org/dataportabilityproject/auth/google/GoogleAuthServiceExtension.java]: authDataGenerator = makeAuthDataGenerator ? 
That or make makeAuthDataGenerator return void?

File: extensions/auth/portability-auth-google/src/main/java/org/dataportabilityproject/auth/google/GoogleAuthServiceExtension.java [hunk 5]
```diff
@@ -0,0 +1,92 @@
...
+
+  public GoogleAuthServiceExtension() {}
+
+  public String getServiceId() {
+    return "google";
+  }
+
+  public AuthDataGenerator getAuthDataGenerator(String transferDataType, AuthMode mode) {
+    // TODO Create auth data generator for a given mode, usually means a different scope
+    if (DEFAULT_AUTH_GENERATOR_SERVICES.contains(transferDataType)) {
```
Comment (seehamrun) [inline extensions/auth/portability-auth-google/src/main/java/org/dataportabilityproject/auth/google/GoogleAuthServiceExtension.java]: we should check here that this extension was initialized and authDataGenerator != null

File: extensions/auth/portability-auth-google/src/main/java/org/dataportabilityproject/auth/google/GoogleAuthServiceExtension.java [hunk 6]
```diff
@@ -0,0 +1,92 @@
...
+import org.dataportabilityproject.api.launcher.ExtensionContext;
+import org.dataportabilityproject.spi.cloud.storage.AppCredentialStore;
+import org.dataportabilityproject.spi.gateway.auth.AuthDataGenerator;
+import org.dataportabilityproject.spi.gateway.auth.AuthServiceProviderRegistry.AuthMode;
+import org.dataportabilityproject.spi.gateway.auth.extension.AuthServiceExtension;
+import org.dataportabilityproject.types.transfer.auth.AppCredentials;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/** */
```
Comment (seehamrun) [inline extensions/auth/portability-auth-google/src/main/java/org/dataportabilityproject/auth/google/GoogleAuthServiceExtension.java]: class comment

File: extensions/auth/portability-auth-google/src/main/java/org/dataportabilityproject/auth/google/GoogleAuthDataGenerator.java (hunk 1 above)
Comment (seehamrun) [reply extensions/auth/portability-auth-google/src/main/java/org/dataportabilityproject/auth/google/GoogleAuthDataGenerator.java]: just noticed that you pass in the id and secret from the AppCredential. It probably doesnt matter but can we change this or the MSAuthExtension to be consistent? 

https://github.com/google/data-transfer-project/blob/125cfb57b8380aad21b6bff0afe62855c4f70501/extensions/auth/portability-auth-microsoft/src/main/java/org/dataportabilityproject/auth/microsoft/MicrosoftAuthDataGenerator.java#L49

Comment (holachuy) [reply extensions/auth/portability-auth-google/src/main/java/org/dataportabilityproject/auth/google/GoogleAuthDataGenerator.java]: Cool, let's refactor in a follow up PR if we want it.

File: extensions/auth/portability-auth-google/src/main/java/org/dataportabilityproject/auth/google/GoogleAuthServiceExtension.java (hunk 5 above)
Comment (holachuy) [reply extensions/auth/portability-auth-google/src/main/java/org/dataportabilityproject/auth/google/GoogleAuthServiceExtension.java]: Funny, I had made that same comment on the MS PR :)

File: extensions/auth/portability-auth-google/src/main/java/org/dataportabilityproject/auth/google/GoogleAuthServiceExtension.java (hunk 3 above)
Comment (holachuy) [reply extensions/auth/portability-auth-google/src/main/java/org/dataportabilityproject/auth/google/GoogleAuthServiceExtension.java]: Done. Good catch!

File: extensions/auth/portability-auth-google/src/main/java/org/dataportabilityproject/auth/google/GoogleAuthServiceExtension.java (hunk 4 above)
Comment (holachuy) [reply extensions/auth/portability-auth-google/src/main/java/org/dataportabilityproject/auth/google/GoogleAuthServiceExtension.java]: Sorry, didn't clean that up. This should only be created once.

File: settings.gradle (hunk 2 above)
Comment (holachuy) [reply settings.gradle]: Good to know.

**Pull request merged**